File: read-only-viewer/rs/src/transcripts.rs
//! Reading JSONL transcripts off disk.
//!
//! Port of `alcove/transcripts.py`. The two traps it guards are unchanged: a
//! tail read cuts the first record mid-line, and file order is not chronology
//! because compaction rewrites transcripts with repeated, out-of-order blocks.

use std::fs::File;
use std::io::{Read, Seek, SeekFrom};
use std::path::Path;

use serde_json::Value;

pub const TAIL_BYTES: u64 = 1 << 20;
pub const TAIL_LINES: usize = 4000;
const HEAD_BYTES: usize = 1 << 16;

/// Parse JSONL, skipping records that do not parse or are not objects.
///
/// A malformed line is skipped rather than fatal: transcripts are appended to
/// live, so the final line is routinely a partial write.
fn parse(data: &[u8], limit: Option<usize>) -> Vec<Value> {
    let text = String::from_utf8_lossy(data);
    let lines: Vec<&str> = text.lines().collect();
    let start = match limit {
        Some(n) if lines.len() > n => lines.len() - n,
        _ => 0,
    };
    lines[start..]
        .iter()
        .filter(|l| !l.trim().is_empty())
        .filter_map(|l| serde_json::from_str::<Value>(l).ok())
        .filter(|v| v.is_object())
        .collect()
}

/// Parsed objects from the TAIL of a transcript.
pub fn tail_events(path: &Path) -> Vec<Value> {
    let mut file = match File::open(path) {
        Ok(f) => f,
        Err(_) => return Vec::new(),
    };
    let size = match file.metadata() {
        Ok(m) => m.len(),
        Err(_) => return Vec::new(),
    };
    let mut buf = Vec::new();
    if size <= TAIL_BYTES {
        if file.read_to_end(&mut buf).is_err() {
            return Vec::new();
        }
    } else {
        if file.seek(SeekFrom::End(-(TAIL_BYTES as i64))).is_err() {
            return Vec::new();
        }
        if file.read_to_end(&mut buf).is_err() {
            return Vec::new();
        }
        // First line is almost certainly cut mid-record.
        if let Some(pos) = buf.iter().position(|&b| b == b'\n') {
            buf.drain(..=pos);
        }
    }
    parse(&buf, Some(TAIL_LINES))
}

/// Parsed objects from the START of a transcript.
///
/// Identity lives in the first record — Codex writes `session_meta` as line
/// one. A tail-only read silently loses identity on any file bigger than the
/// tail window, which reads as "this session does not exist".
pub fn head_events(path: &Path) -> Vec<Value> {
    let mut file = match File::open(path) {
        Ok(f) => f,
        Err(_) => return Vec::new(),
    };
    let mut buf = vec![0u8; HEAD_BYTES];
    let n = match file.read(&mut buf) {
        Ok(n) => n,
        Err(_) => return Vec::new(),
    };
    buf.truncate(n);
    // Drop a trailing partial line so it does not fail to parse.
    if let Some(pos) = buf.iter().rposition(|&b| b == b'\n') {
        buf.truncate(pos);
    }
    parse(&buf, None)
}

/// Dedupe by uuid (first wins) and sort by a string timestamp field.
///
/// Must stay a STABLE sort on the raw string: the Python implementation sorts
/// `str(e.get(key) or "")` with a stable sort, and the equivalence gate compares
/// the resulting order.
pub fn chronological(events: Vec<Value>, key: &str) -> Vec<Value> {
    let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut unique: Vec<Value> = Vec::with_capacity(events.len());
    for event in events {
        if let Some(uuid) = event.get("uuid").and_then(Value::as_str) {
            if !seen.insert(uuid.to_string()) {
                continue;
            }
        }
        unique.push(event);
    }
    unique.sort_by(|a, b| {
        let ka = a.get(key).and_then(Value::as_str).unwrap_or("");
        let kb = b.get(key).and_then(Value::as_str).unwrap_or("");
        ka.cmp(kb)
    });
    unique
}

pub fn file_size(path: &Path) -> u64 {
    path.metadata().map(|m| m.len()).unwrap_or(0)
}

File: read-only-viewer/reference/alcove/transcripts.py
"""Reading JSONL transcripts off disk.

Two traps live here, and both produce confidently wrong output rather than an
error, so they are handled once in this module rather than at each call site.
"""

from __future__ import annotations

import json
import os
import time
from pathlib import Path
from typing import Any

from . import config


def _parse(data: bytes, limit: int | None) -> list[dict[str, Any]]:
    lines = data.decode("utf-8", errors="replace").splitlines()
    if limit is not None:
        lines = lines[-limit:]
    out = []
    for line in lines:
        if not line.strip():
            continue
        try:
            event = json.loads(line)
        except (json.JSONDecodeError, ValueError):
            continue
        if isinstance(event, dict):
            out.append(event)
    return out


def tail_events(path: Path) -> list[dict[str, Any]]:
    """Parsed JSONL objects from the tail of a transcript."""
    try:
        size = path.stat().st_size
    except OSError:
        return []
    try:
        with path.open("rb") as handle:
            if size <= config.TAIL_BYTES:
                data = handle.read()
            else:
                handle.seek(-config.TAIL_BYTES, os.SEEK_END)
                data = handle.read()
                # First line is almost certainly cut mid-record.
                data = data.split(b"\n", 1)[1] if b"\n" in data else data
    except OSError:
        return []
    return _parse(data, config.TAIL_LINES)


def head_events(path: Path, limit_bytes: int = 1 << 16) -> list[dict[str, Any]]:
    """Parsed objects from the START of a transcript.

    Identity lives in the first record — Codex writes `session_meta` (thread id,
    cwd, and the `parent_thread_id` that marks a spawned agent) as line one. A
    tail-only read silently loses identity on any file bigger than the tail
    window, which reads as "this session does not exist" rather than as an error.
    """
    try:
        with path.open("rb") as handle:
            data = handle.read(limit_bytes)
    except OSError:
        return []
    # Drop a trailing partial line so it does not fail to parse.
    if b"\n" in data:
        data = data.rsplit(b"\n", 1)[0]
    return _parse(data, None)


def chronological(events: list[dict[str, Any]],
                  key: str = "timestamp") -> list[dict[str, Any]]:
    """Dedupe by uuid and sort by timestamp.

    Compaction rewrites transcripts with repeated, out-of-order blocks, so raw
    file order is not chronology — scanning it directly reports model switches
    that run backwards in time.
    """
    seen: set[str] = set()
    unique = []
    for event in events:
        uuid = event.get("uuid")
        if isinstance(uuid, str):
            if uuid in seen:
                continue
            seen.add(uuid)
        unique.append(event)
    unique.sort(key=lambda e: str(e.get(key) or ""))
    return unique


def mtime_age(path: Path) -> float | None:
    try:
        return time.time() - path.stat().st_mtime
    except OSError:
        return None


def file_size(path: Path) -> int:
    try:
        return path.stat().st_size
    except OSError:
        return 0
